### DISCUSSION TITLE
Drum-tec Diabolo snare rim triggers too easily

### DISCUSSION THREAD

**Discussion by thijstriemstra** [corrados/edrumulus#30] [General]
I'm testing the Drum-tec Diabolo snare with an ESP32 and this snare config:

| Pad | Type | MIDI note | GPIO |
| --- | --- | --- | --- |
| Drum-tec Diabolo | Snare | 38, 40 | 36, 35 |

and it seems the rim is being triggered way too easily. 

Here's a video and below that a screenshot with the associated Ableton midi notes, where you can see the velocities. The notes in blue are rim.

https://user-images.githubusercontent.com/305679/147894102-e08ca822-1511-46eb-b745-ac32db7dce81.mp4

![drumtec-snare-velocity](https://user-images.githubusercontent.com/305679/147894125-6a020c42-6a84-4a15-9376-4ee14190894d.jpg)

**Comment by corrados** [comment]
See my [following comment](https://github.com/corrados/edrumulus/discussions/29#discussioncomment-1900825) to adjust the Edrumulus pad parameters correctly.

**Comment by thijstriemstra** [comment]
I adjust the pad parameters like this:

```c++
  // SNARE (Drum-tec Diabolo)
  edrumulus.set_pad_type           ( 0, Edrumulus::PD120 );
  edrumulus.set_rim_shot_is_used   ( 0, true );
  edrumulus.set_pos_sense_is_used  ( 0, true );
  edrumulus.set_pos_threshold      ( 0, 4 );
  edrumulus.set_pos_sensitivity    ( 0, 17 );
  edrumulus.set_rim_shot_treshold  ( 0, 25 );
```

And this is the result (ps. im not a drummer! [yet]):


https://user-images.githubusercontent.com/305679/148126709-4d7e8bb4-148d-4bca-83ca-d40ebcb428b2.mp4


As you can hear the rimshot is quite bad and I wasn't able to find the good parameters for it. It sometimes even does double hits and overall it's just really noisy/sensitive. I can simply disable it, I think this would fix a lot of issues with it randomly triggering, not triggering at all etc. 

```c++
edrumulus.set_rim_shot_is_used   ( 0, false );
```

Also, I couldn't get the start of positional sensing MIDI note 16 to get really close to the middle after trying many combinations of params. The closest to the center I could get was a few cm under the logo on the mesh, and from there that area goes from 1 towards 127 at the edge of mesh. Having it even closer to the middle, and making the range larger would also be nice.

**Reply by corrados** [reply]
> As you can hear the rimshot is quite bad and I wasn't able to find the good parameters for it.

What you play is not a rimshot. A rimshot is the stroke method to hit rim of the snare and snare at the same time. See, e.g., https://www.facebook.com/permalink.php?story_fbid=137495667978256&id=103858198008670.

You are hitting the rim directly which is not (yet) supported by Edrumulus.

> The closest to the center I could get was a few cm under the logo on the mesh, and from there that area goes from 1 towards 127 at the edge of mesh. 

To debug this problem it would be good to have an audio recording of the piezo signal. This signal can be recorded by plugging in the piezo cable in a mic/line in of a mixer and record a wave file. Ideally, you would strike the mesh head from the far left and move to the right edge and do this multiple times. Also it would be good to have this with low velocity strikes and mid/high velocity strikes.

**Reply by thijstriemstra** [reply]
> You are hitting the rim directly which is 

Rim, rimshot, sounds the same to me 😄 

> not (yet) supported by Edrumulus.

Good to know!

> To debug this problem it would be good to have an audio recording of the piezo signal. 

Here's a zip file with 6 recordings. The files ending with 1 are from left to right, 2 are right to left.

[drumtec-diabolo-snare.zip](https://github.com/corrados/edrumulus/files/7821176/drumtec-diabolo-snare.zip)

**Reply by corrados** [reply]
Thanks for the recordings. Unfortunately, it seems that you have recorded the rim piezo and not the piezo which is located in the middle of the mesh head. Can you please check that you have recorded the correct channel of your pad?

**Reply by thijstriemstra** [reply]
I recorded with a stereo jack cable (which is also used when using edrumulus snare) but I can use a mono jack if that is what you mean?

**Reply by corrados** [reply]
The wave files in your zip are mono. The Drumtec pad is stereo (i.e. it has two piezos). I either need the piezo signal from the mesh head in a mono file or maybe record both piezo signals in a stereo wave file.

**Reply by thijstriemstra** [reply]
> The wave files in your zip are mono.

I see. I re-recorded using a cable like the one below (with extra adapter so mini jack fits in the snare).

And I also included a wav that hits the rim all way around clockwise, starting at the top center of snare.

Also added #34 so you can see params I used.

[drumtec-diabolo-snare2.zip](https://github.com/corrados/edrumulus/files/7822522/drumtec-diabolo-snare2.zip)

![GREATLINK-3-5mm-TRS-Stereo-Male-naar-Dual-XLR-Male-Splitter-Patch-Kabel-Ongebalanceerde-Mini-Jack](https://user-images.githubusercontent.com/305679/148398762-4b5511ea-7abe-4e72-a769-c8b0ff537efe.jpg)

**Reply by corrados** [reply]
Thank you again. I have added your signal to the repo (I assumed that this is ok for you).

This Diabolo pad signal looks so different from all I have seen so far. The bad news is that the positional sensing algorithm in Edrumulus seems not to work for this pad. I have no idea why because it has a center piezo and on the drum-tec homepage they claim that they are 100% compatible to Roland drum modules. So, in principle, it should work out of the box. I have seen videos where the Diabolo pad is used with a Roland drum module and the positional sensing is working perfectly. Very strange...

More investigations are needed...

**Reply by thijstriemstra** [reply]
> (I assumed that this is ok for you).

Yep, thanks.

> So, in principle, it should work out of the box. I have seen videos where the Diabolo pad is used with a Roland drum module and the positional sensing is working perfectly.

I can see midi note 16 and tune it, and my drummer friend and I confirmed we can hear the positional sensing. We've never hooked it up to a Roland module that supports positional sensing, only a Roland Octopad SPD-30 which doesn't support it. So we have nothing to compare it to except for SPD-30 and it sounds 100x better when it comes to sensitivity/sensing etc..

> More investigations are needed...

Sounds good! If you need more recordings etc, let me know.

Btw, I only tested it with Addictive Drums 2 till last night; but when we hooked it up to Superior Drummer 3, boy, that lifts the whole thing to the next level.

**Reply by corrados** [reply]
I assume it is a 12 inch Diabolo pad, right?

I have added experimental support for this pad now: https://github.com/corrados/edrumulus/commit/0300ec69981d337b663c3824b10a06fd13e65bfd

I see a strong "hot spot" effect and the positional sensing does not really work with the current algorithm. Would be interesting to know if the eDRUMin has correct support for positional sensing for this pad type.

BTW: I noticed in your videos that you have a high latency. For real-time usage I recommend to set the audio buffer in your DAW to a very low value like 64 or 32.

**Reply by thijstriemstra** [reply]
> I assume it is a 12 inch Diabolo pad, right?

Think so.

> I have added experimental support for this pad now

Nice! Anything I can test and how?

> Would be interesting to know if the eDRUMin has correct support for positional sensing for this pad type.

I don't own one but I can check their documentation perhaps.

> I noticed in your videos that you have a high latency. For real-time usage I recommend to set the audio buffer in your DAW to a very low value like 64 or 32.

I hardly experience any latency, quite the opposite. The video is recorded with an iphone 12, a couple of studio monitors for AD2 in the background, and the hits on the pads are quite loud, hard to say from a video if there's latency imo. Nevertheless, I am using 512, you're right and should try a buffer size of 64 and 32 samples as well. Drums we make here is always quantized, the latency is less of an issue then, and my drummer friend is very happy with results at 512. Also, jamming along (multiple) [live] midi vst instruments (e.g. serum, anything native instruments or izotope for example) requires > 64 to avoid popping.. But will give it a try!

SoI'll keep supplying any.wav files at 512 if that makes sense?

**Reply by thijstriemstra** [reply]
> I don't own one but I can check their documentation perhaps.

From their [forum](https://www.audiofront.net/forum/viewtopic.php?f=29&t=611):

```
drum-tec 13” Diabolo Hand Hammered Snare Pad works 100%, with Positional Sensing.

These are listed in the manual so considered officially supported but users have provided configurations that have worked... so good information to have.
```


Presumably the same pos. sensing tech?

**Reply by corrados** [reply]
> >  I have added experimental support for this pad now

> Nice! Anything I can test and how?

Yes, you can merge from main if you like.

> SoI'll keep supplying any.wav files at 512 if that makes sense?

For the recordings you should has a large buffer size. Just for real-time usage, a smaller buffer should be used.

> drum-tec 13” Diabolo Hand Hammered Snare Pad works 100%, with Positional Sensing

Your Diabolo wodden snare [looks like this in the inside](http://www.rhythmz.de/img/diabolo4.jpg). But the Hammered Snare looks different when it comes to the piezo construction [as you can see here](https://www.drumforum.org/attachments/c25256ef-9c7d-449d-b832-0da8e9b138dd-jpeg.472863/). I am pretty sure they have different resonance behavior when it comes to triggering since the Hammered Snare piezo construction looks much more solid.

**Reply by thijstriemstra** [reply]
> For the recordings you should has a large buffer size.

Ok.

> Yes, you can merge from main if you like.

See #37
